cpp/src/c_api/uniform_neighbor_sampling.cpp

```diff
@@ -222,6 +222,80 @@ extern "C" cugraph_type_erased_host_array_view_t* cugraph_sample_result_get_coun
   return reinterpret_cast<cugraph_type_erased_host_array_view_t*>(internal_pointer->count_->view());
 }
 
+extern "C" cugraph_error_code_t cugraph_sample_result_create(
+  const cugraph_resource_handle_t* handle,
+  const cugraph_type_erased_host_array_view_t* srcs,
+  const cugraph_type_erased_host_array_view_t* dsts,
+  const cugraph_type_erased_host_array_view_t* weights,
+  const cugraph_type_erased_host_array_view_t* counts,
+  cugraph_sample_result_t** result,
+  cugraph_error_t** error)
+{
+  *result = nullptr;
+  *error  = nullptr;
+  size_t n_bytes{0};
+
+  try {
+    if (!handle) {
+      *error = reinterpret_cast<cugraph_error_t*>(
+        new cugraph::c_api::cugraph_error_t{"invalid resource handle"});
+      return CUGRAPH_INVALID_HANDLE;
+    }
+
+    auto p_handle = reinterpret_cast<cugraph::c_api::cugraph_resource_handle_t const*>(handle);
+    auto host_srcs =
+      reinterpret_cast<const cugraph::c_api::cugraph_type_erased_host_array_view_t*>(srcs);
+    auto new_device_srcs = new cugraph::c_api::cugraph_type_erased_device_array_t(
+      host_srcs->size_, host_srcs->num_bytes_, host_srcs->type_, p_handle->handle_->get_stream());
+    raft::update_device(reinterpret_cast<byte_t*>(new_device_srcs->data_.data()),
```

Comment:
If you're not going to do type dispatching, consider the following.

You could actually call `cugraph_type_erased_device_array_create` and `cugraph_type_erased_device_array_view_copy_from_host` and leverage some of the already written code.  This also gets you around having to do so much type casting.  This might be a bit cleaner (untested):
```
cugraph_type_erased_device_array_t *new_src{nullptr};
err = cugraph_type_erased_device_array_create(handle, host_srcs->size_, host_srcs_->type_, &new_src, error);
// check the value of err, return if it's not CUGRAPH_SUCCESS
auto new_device_sources_view = cugraph_type_erased_device_array_view(new_device_sources);
cugraph_type_erased_device_array_view_copy_from_host(
  handle,
  & new_device_sources_view,
  host_srcs->data_->as_type<byte_t>(),
  error);
// check the value of err, return if it's not CUGRAPH_SUCCESS
```

You might not need, then, to create p_handle and a few of the other things you're doing.

Reply:
I did this in the most recent commit, and I also used cugraph_type_erased_device_array_create_from_view()` from #2712 , based on our discussion offline.